File: fsync_latency/test_fsync_latency.c
#define _GNU_SOURCE  // To enable O_DIRECT
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <string.h>
#include <time.h>

#define FILE_NAME "testfile"
#define BUFFER_SIZE 512

void test_sync(int rounds) {
    printf("Start testing %d rounds...\n", rounds);
    printf("getpagesize() = %d\n", getpagesize());

    // Open a file
    int fd = open(FILE_NAME, O_RDWR | O_CREAT | O_DIRECT, S_IRUSR | S_IWUSR);
    if (fd == -1) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    // Extend the file size to the buffer size
    if (ftruncate(fd, BUFFER_SIZE) == -1) {
        perror("Error extending file");
        close(fd);
        exit(EXIT_FAILURE);
    }

    char *buffer = mmap(NULL, BUFFER_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (buffer == MAP_FAILED) {
        perror("Error mapping file to memory");
        close(fd);
        exit(EXIT_FAILURE);
    }

    int written = 0;
    for (int i = 0; i < rounds; ++i) {
        buffer[i % BUFFER_SIZE] = rand() % 256;
#if 0
        msync(buffer, BUFFER_SIZE, MS_SYNC);  // Synchronize changes to the file
#else
        lseek(fd, 0, SEEK_SET);
        ssize_t res = write(fd, buffer, BUFFER_SIZE);
        if (res == -1) {
            perror("Error writing to file");
            close(fd);
            exit(EXIT_FAILURE);
        }
        written += res;
        fsync(fd);
#endif
    }

    printf("Written %d bytes\n", written);

    // Unmap the memory
    if (munmap(buffer, BUFFER_SIZE) == -1) {
        perror("Error unmapping file from memory");
        close(fd);
        exit(EXIT_FAILURE);
    }

    // Close opened file
    if (close(fd) == -1) {
        perror("Error closing file");
        exit(EXIT_FAILURE);
    }

    // Remove the file
    if (remove(FILE_NAME) == -1) {
        perror("Error removing file");
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
      printf("Usage: time %s <rounds>\n", argv[0]);
      printf("rounds - a number of fsync calls\n");
      exit(EXIT_FAILURE);
    }

    srand((unsigned int)time(NULL)); // Seed for random number generation

    int rounds = atoi(argv[1]);
    test_sync(rounds);

    return 0;
}

File: tools/test_fsync_latency.py
#!/usr/bin/python

import os, sys, mmap, random
#import cProfile

FILE_NAME="testfile"
ROUNDS=1000
BUFFER_SIZE=512

if len(sys.argv) < 2:
   print(f"Usage: time python {sys.argv[0]} [rounds]")
   print("rounds - a number of fsync calls (default=1000)")
   print("\nReturns e.g.:")
   print("# real  0m6,334s")
   print("Where \"real\" shows latency = 6,334s/1000 rounds = 6,334 ms")
   sys.exit(1)

def test_sync():
   WRITTEN=0
   ROUNDS=int(sys.argv[1])
   print(f"Start testing {ROUNDS} rounds...")

   # Open a file
   fd = os.open(FILE_NAME, os.O_RDWR|os.O_CREAT|os.O_DIRECT)
   os.truncate(fd, BUFFER_SIZE)

   m = mmap.mmap(-1, BUFFER_SIZE)

   for i in range (0, ROUNDS):
      os.lseek(fd,os.SEEK_SET,0)
      m[i % BUFFER_SIZE] = random.randint(0, 255)
      WRITTEN += os.write(fd, m)
      os.fsync(fd)

   print(f"Written {WRITTEN} bytes...")

   # Close memory-mapped file
   m.close()

   # Close opened file
   os.close(fd)
   os.remove(FILE_NAME)

#cProfile.run("test_sync()", sort='cumulative')
test_sync()
